unified-scheduler-pool/src/sleepless_testing.rs

```diff
@@ -0,0 +1,352 @@
+            thread::{current, JoinHandle, ThreadId},
+        },
+    };
+
+    #[derive(Debug)]
+    struct Progress {
+        _name: String,
+        check_points: Vec<String>,
+        current_check_point: Mutex<String>,
+        condvar: Condvar,
+    }
+
+    #[derive(Debug)]
+    struct JustCreated;
+
+    impl Progress {
+        fn new(check_points: impl Iterator<Item = String>, name: String) -> Self {
+            let initial_check_point = format!("{JustCreated:?}");
+            let check_points = [initial_check_point.clone()]
+                .into_iter()
+                .chain(check_points)
+                .collect::<Vec<_>>();
+            let check_points_set = check_points.iter().collect::<HashSet<_>>();
+            assert_eq!(check_points.len(), check_points_set.len());
+
+            Self {
+                _name: name,
+                check_points,
+                current_check_point: Mutex::new(initial_check_point),
+                condvar: Condvar::new(),
+            }
+        }
+
+        fn change_current_check_point(&self, anchored_check_point: String) {
+            let Some(anchored_index) = self
+                .check_points
+                .iter()
+                .position(|check_point| check_point == &anchored_check_point)
+            else {
+                // Ignore unrecognizable checkpoints...
+                return;
+            };
+
+            let mut current_check_point = self.current_check_point.lock().unwrap();
+
+            let should_change =
+                match anchored_index.cmp(&self.expected_next_index(&current_check_point)) {
+                    Equal => true,
+                    Greater => {
+                        // anchor is one of future check points; block the current thread until
+                        // that happens
+                        current_check_point = self
+                            .condvar
+                            .wait_while(current_check_point, |current_check_point| {
+                                anchored_index != self.expected_next_index(current_check_point)
+                            })
+                            .unwrap();
+                        true
+                    }
+                    // anchor is already observed.
+                    Less => false,
+                };
+
+            if should_change {
+                *current_check_point = anchored_check_point;
+                self.condvar.notify_all();
+            }
+        }
+
+        fn expected_next_index(&self, current_check_point: &String) -> usize {
+            let current_index = self
+                .check_points
+                .iter()
+                .position(|check_point| check_point == current_check_point)
+                .unwrap();
+            current_index.checked_add(1).unwrap()
+        }
+    }
+
+    lazy_static! {
+        static ref THREAD_REGISTRY: Mutex<HashMap<ThreadId, Arc<Progress>>> =
+            Mutex::new(HashMap::new());
+    }
```

Comment:
this is a global. won't it cause contention between all tests using this?

Reply:
yes, I'm leaning towards switching to `thread_local!` impl soon.